docs/map.rst

```diff
@@ -0,0 +1,602 @@
+
+    Map.!? :: Ord k => Map k v -> k -> Maybe v
+    Map.!? m key = ...
+
+:haddock_short:`Data.Map.Strict#lookup` the value corresponding to the given
+``key``, returns ``Nothing`` if the key is not present; the ``!?`` operator
+(*since 0.5.10*) is a flipped version of ``lookup`` and can often be imported
+unqualified.
+
+
+If you want to provide a default value if the key doesn't exist you can do:
+
+::
+
+    import Data.Maybe (fromMaybe)
+
+    -- fromMaybe :: a -> Maybe a -> a
+    fromMaybe defaultValue (lookup k m)
+
+For example::
+
+    import Data.Map.Strict ((!?))
+    import Data.Maybe (fromMaybe)
+
+    Map.lookup 1 Map.empty
+    > Nothing
+
+    Map.lookup 1 (Map.fromList [(1,"one"),(2,"two"),(3,"three")])
+    > Just "one"
+
+    > (Map.fromList [(1,"one"),(2,"two"),(3,"three")]) !? 1
+    > Just "one"
+
+    fromMaybe "?" (Map.empty !? 1)
+    > "?"
+
+    fromMaybe "?" (Map.fromList [(1,"one"), (2,"two"), (3,"three")] !? 1)
+    > "one"
+
+.. WARNING::
+   **DO NOT** Use ``Map.!``. It is partial and throws a runtime error if the key
+   doesn't exist.
+
+Find the minimum/maximum
+""""""""""""""""""""""""
+
+*Since version 0.5.9*
+
+::
+
+    Map.lookupMin, Map.lookupMax :: Map k v -> Maybe (k, v)
+    Map.lookupMin m = ...
+    Map.lookupMax m = ...
+
+:haddock_short:`Data.Map.Strict#lookupMin` and
+:haddock_short:`Data.Map.Strict#lookupMax` respectively return the
+minimum or maximum element of the map ``m``, or ``Nothing`` if the map is empty.
+
+::
+
+    Map.lookupMin Map.empty
+    > Nothing
+
+    Map.lookupMin (Map.fromList [(1,"one"), (2,"two"), (3,"three")])
+    > Just (1,"one")
+
+    Map.lookupMax (Map.fromList [(1,"one"), (2,"two"), (3,"three")])
+    > Just (3,"three")
+
+.. WARNING::
+   **DO NOT** use ``Map.findMin`` or ``Map.findMax``. They are partial and throw
+   a runtime error if the map is empty.
+
+Modification
+^^^^^^^^^^^^
+
+Adding a new entry to a map
+"""""""""""""""""""""""""""
+
+::
+
+    Map.insert :: Ord k => k -> v -> Map k v -> Map k v
+    Map.insert key value m = ...
+
+:haddock_short:`Data.Map.Strict#insert` adds the ``value`` into the map ``m``
+with the given ``key``, replacing the existing value if the key already exists.
+
+::
+
+    Map.insert 1 "one" Map.empty
+    > Map.fromList [(1,"one")]
+
+    Map.insert 4 "four" (Map.fromList [(1,"one"), (2,"two"), (3,"three")])
+    > fromList [(1,"one"),(2,"two"),(3,"three"),(4,"four")]
+
+    Map.insert 1 "uno" (Map.fromList [(1,"one"), (2,"two"), (3,"three")])
+    > fromList [(1,"uno"),(2,"two"),(3,"three")]
+
+
+Removing an entry from a map
+""""""""""""""""""""""""""""
+
+::
+
+    Map.delete :: Ord k => k -> Map k v -> Map k v
+    Map.delete key m = ...
+
+:haddock_short:`Data.Map.Strict#delete` removes the entry with the specified
+``key`` from the map ``m``.  If the key doesn't exist it leaves the map
+unchanged. Remember, maps are immutable so if you delete an entry from a map you
+need to assign the new map to a new variable.
```

Comment:
The same warning would also apply to `insert`, `alter`, `adjust`, etc. I think it would be better to condense all such discussion. Near the top of this thing you introduce `Map` as an immutable data structure. I suggest you explain a bit about what that means there, and then probably never mention it again.

Reply:
Done. Made a similar change for the Set docs.